{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0c9658fa-7fe5-449e-8c7b-5304fe3dca7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pip install aiohttp asyncssh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "967881f2-94f5-4124-8ba5-a0209c9db4b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import asyncio\n",
    "import asyncssh\n",
    "import aiohttp\n",
    "from http.client import HTTPConnection\n",
    "from urllib.parse import urlparse\n",
    "\n",
    "\n",
    "async def site_is_online_async(url, timeout):\n",
    "    \"\"\"Return True if the target URL is online.\n",
    "\n",
    "    Raise an exception otherwise.\n",
    "    \"\"\"\n",
    "    error = Exception(\"unknown scheme\")\n",
    "    \n",
    "    parser = urlparse(url)\n",
    "    host = parser.netloc or parser.path.split(\"/\")[0]\n",
    "    \n",
    "    \n",
    "    if parser.scheme in (\"http\", \"https\"):\n",
    "        async with aiohttp.ClientSession() as session:\n",
    "            try:\n",
    "                await session.head(url, timeout=timeout,ssl=False)\n",
    "                return True\n",
    "            except asyncio.TimeoutError:\n",
    "                error = Exception(\"timed out\")\n",
    "            except (asyncssh.Error) as e:\n",
    "                error = e\n",
    "                \n",
    "    if parser.scheme in (\"ssh\"):\n",
    "        if  parser.port == None : \n",
    "            port=22 \n",
    "        else: \n",
    "            port=parser.port\n",
    "            host=host.split(\":\")[0]\n",
    "\n",
    "        #print(host,port)\n",
    "        \n",
    "        try:\n",
    "            async with await asyncio.wait_for(asyncssh.connect(host=host,port=port,username=\"test\", password=\"test\", known_hosts=None), timeout=timeout) as conn:\n",
    "                print((await conn.run('tty', term_type='ansi')).stdout, end='')\n",
    "                return True\n",
    "        except asyncio.TimeoutError:\n",
    "                error = Exception(\"timed out\")\n",
    "        except (asyncssh.Error) as e:\n",
    "                error = e\n",
    "                \n",
    "    if parser.scheme in (\"postgre\"):                \n",
    "        if  parser.port == None : \n",
    "            port=5432\n",
    "        else: \n",
    "            port=parser.port\n",
    "            host=host.split(\":\")[0]\n",
    "        \n",
    "        coro = asyncio.open_connection(host, port)\n",
    "        # execute the coroutine with a timeout\n",
    "        try:\n",
    "            # open the connection and wait for a moment\n",
    "            _,writer = await asyncio.wait_for(coro, timeout)\n",
    "            # close connection once opened\n",
    "            writer.close()\n",
    "            # indicate the connection can be opened\n",
    "            return True\n",
    "        except asyncio.TimeoutError:\n",
    "                error = Exception(\"timed out\")\n",
    "        except (asyncssh.Error) as e:\n",
    "                error = e\n",
    "                \n",
    "    if parser.scheme in (\"kafka\"):                \n",
    "        if  parser.port == None : \n",
    "            port=9092\n",
    "        else: \n",
    "            port=parser.port\n",
    "            host=host.split(\":\")[0]\n",
    "        \n",
    "        coro = asyncio.open_connection(host, port)\n",
    "        # execute the coroutine with a timeout\n",
    "        try:\n",
    "            # open the connection and wait for a moment\n",
    "            _,writer = await asyncio.wait_for(coro, timeout)\n",
    "            # close connection once opened\n",
    "            writer.close()\n",
    "            # indicate the connection can be opened\n",
    "            return True\n",
    "        except asyncio.TimeoutError:\n",
    "                error = Exception(\"timed out\")\n",
    "        except (asyncssh.Error) as e:\n",
    "                error = e\n",
    "                \n",
    "    raise error\n",
    "    \n",
    "def display_check_result(result, url, error=\"\"):\n",
    "    \"\"\"Display the result of a connectivity check.\"\"\"\n",
    "    print(f'The status of \"{url}\" is:', end=\" \")\n",
    "    if result:\n",
    "        print('\"Online!\" üëç')\n",
    "    else:\n",
    "        print(f'\"Offline?\" üëé \\n  Error: \"{error}\"')\n",
    "\n",
    "\n",
    "async def _asynchronous_check(urls,timeout):\n",
    "    async def _check(url,timeout):\n",
    "        error = \"\"\n",
    "        try:\n",
    "            result = await site_is_online_async(url,timeout)\n",
    "        except Exception as e:\n",
    "            result = False\n",
    "            error = str(e)\n",
    "        display_check_result(result, url, error)\n",
    "\n",
    "    await asyncio.gather(*(_check(url,timeout) for url in urls))\n",
    "\n",
    "def check(urls,timeout=5):\n",
    "    try:\n",
    "        loop = asyncio.get_running_loop()\n",
    "    except RuntimeError:  # 'RuntimeError: There is no current event loop...'\n",
    "        loop = None\n",
    "\n",
    "    if loop and loop.is_running():\n",
    "        print('Async event loop already running. Adding coroutine to the event loop.')\n",
    "        tsk = loop.create_task(_asynchronous_check(urls,timeout))\n",
    "        # ^-- https://docs.python.org/3/library/asyncio-task.html#task-object\n",
    "        # Optionally, a callback function can be executed when the coroutine completes\n",
    "        tsk.add_done_callback(\n",
    "            lambda t: print(f'Task done with result={t.result()}  << return val of main()'))\n",
    "    else:\n",
    "        print('Starting new event loop')\n",
    "        result = asyncio.run(_asynchronous_check(urls,timeout))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7f9bbdd1-7808-4d7a-b195-a59680cc1cca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Async event loop already running. Adding coroutine to the event loop.\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "time.sleep(1)\n",
    "check(timeout=3,urls=[])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "66bbf506-df71-4585-a67b-a310fa777727",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Task done with result=None  << return val of main()\n",
      "Async event loop already running. Adding coroutine to the event loop.\n",
      "The status of \"https://scanme.nmap.org\" is: \"Offline?\" üëé \n",
      "  Error: \"Cannot connect to host scanme.nmap.org:443 ssl:False [Connect call failed ('45.33.32.156', 443)]\"\n",
      "The status of \"kafka://scanme.nmap.org:9093\" is: \"Offline?\" üëé \n",
      "  Error: \"Multiple exceptions: [Errno 111] Connect call failed ('45.33.32.156', 9093), [Errno 99] Cannot assign requested address\"\n",
      "The status of \"kafka://scanme.nmap.org\" is: \"Offline?\" üëé \n",
      "  Error: \"Multiple exceptions: [Errno 111] Connect call failed ('45.33.32.156', 9092), [Errno 99] Cannot assign requested address\"\n",
      "The status of \"postgre://scanme.nmap.org\" is: \"Offline?\" üëé \n",
      "  Error: \"Multiple exceptions: [Errno 111] Connect call failed ('45.33.32.156', 5432), [Errno 99] Cannot assign requested address\"\n",
      "The status of \"ssh://scanme.nmap.org:2222\" is: \"Offline?\" üëé \n",
      "  Error: \"Multiple exceptions: [Errno 111] Connect call failed ('45.33.32.156', 2222), [Errno 99] Cannot assign requested address\"\n",
      "The status of \"kafka://scanme.nmap.org:9092\" is: \"Offline?\" üëé \n",
      "  Error: \"Multiple exceptions: [Errno 111] Connect call failed ('45.33.32.156', 9092), [Errno 99] Cannot assign requested address\"\n",
      "The status of \"http://scanme.nmap.org:80\" is: \"Online!\" üëç\n",
      "The status of \"http://scanme.nmap.org:8080\" is: \"Offline?\" üëé \n",
      "  Error: \"Cannot connect to host scanme.nmap.org:8080 ssl:False [Connect call failed ('45.33.32.156', 8080)]\"\n",
      "The status of \"https://scanme.nmap.org:443\" is: \"Offline?\" üëé \n",
      "  Error: \"Cannot connect to host scanme.nmap.org:443 ssl:False [Connect call failed ('45.33.32.156', 443)]\"\n",
      "The status of \"http://scanme.nmap.org\" is: \"Online!\" üëç\n",
      "The status of \"http://scanme.nmap.org/\" is: \"Online!\" üëç\n",
      "The status of \"ssh://scanme.nmap.org\" is: \"Offline?\" üëé \n",
      "  Error: \"Permission denied\"\n",
      "The status of \"ssh://scanme.nmap.org:22\" is: \"Offline?\" üëé \n",
      "  Error: \"Permission denied\"\n",
      "The status of \"ssh://scanme.nmap.org\" is: \"Offline?\" üëé \n",
      "  Error: \"Permission denied\"\n",
      "Task done with result=None  << return val of main()\n"
     ]
    }
   ],
   "source": [
    "check(timeout=5, urls=[\"kafka://scanme.nmap.org\",\n",
    "                       \"kafka://scanme.nmap.org:9092\",\n",
    "                       \"kafka://scanme.nmap.org:9093\",\n",
    "                       \"postgre://scanme.nmap.org\",\n",
    "                       \"http://scanme.nmap.org\",\n",
    "                       \"http://scanme.nmap.org/\",\n",
    "                       \"http://scanme.nmap.org:80\",\n",
    "                       \"http://scanme.nmap.org:8080\",\n",
    "                       \"https://scanme.nmap.org\",\n",
    "                       \"https://scanme.nmap.org:443\",\n",
    "                       \"ssh://scanme.nmap.org\",\n",
    "                       \"ssh://scanme.nmap.org\",\n",
    "                       \"ssh://scanme.nmap.org:22\",\n",
    "                       \"ssh://scanme.nmap.org:2222\"\n",
    "                      ]\n",
    "     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aaf9f4e5-01cf-4c0f-bec5-60712092e589",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
